lib/src/geometry/line_segment.dart

```diff
@@ -0,0 +1,104 @@
+import 'line.dart';
+import '../../extensions.dart';
+
+class LineSegment {
+  final Vector2 from;
+  final Vector2 to;
+
+  LineSegment(this.from, this.to);
+
+  /// If the line segment is outside of the hitbox
+  bool isOutside(List<Vector2> hitbox) {
+    for (int i = 0; i < hitbox.length; ++i) {
+      if (intersections(LineSegment(hitbox[i], hitbox[(i + 1) % hitbox.length]))
+          .isNotEmpty) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /// Returns an empty list if there are no intersections
+  /// If the segment overlap (not cross) the middle of the intersecting part is
+  /// the result
+  List<Vector2> intersections(LineSegment otherSegment) {
+    final result = toLine().intersections(otherSegment.toLine());
+    if (result.isNotEmpty) {
+      // The lines are not parallel
+      final intersection = result.first;
+      if (containsPoint(intersection) &&
+          otherSegment.containsPoint(intersection)) {
+        // The intersection point is on both line segments
+        return result;
+      }
+    } else {
+      // In here we know that the lines are parallel
+      final overlaps = ({
+        from: otherSegment.containsPoint(from),
+        to: otherSegment.containsPoint(to),
+        otherSegment.from: containsPoint(otherSegment.from),
+        otherSegment.to: containsPoint(otherSegment.to),
+      }..removeWhere((_key, onSegment) => !onSegment))
+          .keys
+          .toSet();
+      if (overlaps.isNotEmpty) {
+        return [
+          overlaps.fold<Vector2>(
+                Vector2.zero(),
+                (sum, point) => sum + point,
+              ) /
+              overlaps.length.toDouble(),
+        ];
+      }
+    }
+    return [];
+  }
+
+  bool containsPoint(Vector2 point) {
+    const epsilon = 0.00001;
+    final delta = to - from;
+    final crossProduct =
+        (point.y - from.y) * delta.x - (point.x - from.x) * delta.y;
+
+    // compare versus epsilon for floating point values
+    if (crossProduct.abs() > epsilon) {
+      return false;
+    }
+
+    final dotProduct =
+        (point.x - from.x) * delta.x + (point.y - from.y) * delta.y;
```

Comment:
ah we should definitely add dot and cross products to the Vector2 class!

Reply:
They already exists as `.dot` and `.cross`.
Do you know a good way to re-write this to get `delta` incorporated into the expression if I'd use those ones instead?

Reply:
that is not exactly a dot/cross product now that I think about it. cross product only exists for vectors 3+d. and that dot product seems off. that expressions seems just to be (point - from) * delta. I will do some digging if there is a correlation there, but it should be 2 3d vectors not 3 2d vectors

Reply:
ah this is (point - from) dot delta. dot product is just our regular multiplication in 2d.
now for the cross product I am not sure why yet

Reply:
yeah the first one is doing vx * uy - ux * vy which looks like the magnitude of 3d cross product formula but isn't really. however that is what our implementation does, so if you do (point - from) cross delta it should be equivalent to that expression. it's a bit of a hack to extend the def to 2d because you can only get the magnitude, the cross product is a vectorial operation

Reply:
Let's leave these for now and maybe improve on them later if we come up with a good way to do that then?